### PR TITLE
[INFRA/RULES] Added Real valued rule evaluation

### PR DESCRIPTION
**Pull request by Neelesh99**
# Summary

Real valued rule evaluation has been added with new aux code 3. Moved basic rule evaluator out of client into rules package

## Additional Information

New tests added.

## Test Plan

- Additional tests added.

- `go test ./...`
```
?       github.com/SOMAS2020/SOMAS2020  [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team1   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team2   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team3   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team4   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team5   [no test files]
?       github.com/SOMAS2020/SOMAS2020/internal/clients/team6   [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/common  (cached) [no tests to run]
?       github.com/SOMAS2020/SOMAS2020/internal/common/config   [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/common/rules    0.121s
?       github.com/SOMAS2020/SOMAS2020/internal/common/shared   [no test files]
ok      github.com/SOMAS2020/SOMAS2020/internal/server  (cached)
?       github.com/SOMAS2020/SOMAS2020/pkg/testutils    [no test files]

```

- `go run .`
```
2020/12/12 19:08:01 [SERVER]: start runTurn
2020/12/12 19:08:01 [SERVER]: TURN: 1, Season: 1
2020/12/12 19:08:01 [SERVER]: start updateIsland
2020/12/12 19:08:01 [Team5]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/12 19:08:01 [Team6]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/12 19:08:01 [Team1]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/12 19:08:01 [Team2]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/12 19:08:01 [Team3]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/12 19:08:01 [Team4]: Received game state update: {1 1 map[Team1:{100 Alive 0} Team2:{100 Alive 0} Team3:{100 Alive 0} Team4:{100 Alive 0} Team5:{100 Alive 0} Team6:{100 Alive 0}]}
2020/12/12 19:08:01 [SERVER]: finish updateIsland
2020/12/12 19:08:01 [SERVER]: start runOrgs
2020/12/12 19:08:01 [SERVER]: start runIITO
2020/12/12 19:08:01 [SERVER]: finish runIITO
2020/12/12 19:08:01 [SERVER]: start runIIFO
2020/12/12 19:08:01 [SERVER]: finish runIIFO
2020/12/12 19:08:01 [SERVER]: start runIITO
2020/12/12 19:08:01 [SERVER]: finish runIITO
2020/12/12 19:08:01 [SERVER]: finish runOrgs
2020/12/12 19:08:01 [SERVER]: start getEndOfTurnActions
2020/12/12 19:08:01 [Team6]: EndOfTurnActions
2020/12/12 19:08:01 [Team1]: EndOfTurnActions
2020/12/12 19:08:01 [Team2]: EndOfTurnActions
2020/12/12 19:08:01 [Team3]: EndOfTurnActions
2020/12/12 19:08:01 [Team4]: EndOfTurnActions
2020/12/12 19:08:01 [Team5]: EndOfTurnActions
2020/12/12 19:08:01 [SERVER]: finish getEndOfTurnActions
2020/12/12 19:08:01 [GAMESTATE]: start DispatchActions
2020/12/12 19:08:01 [GAMESTATE]: finish DispatchActions
2020/12/12 19:08:01 [SERVER]: start endOfTurn
2020/12/12 19:08:01 [SERVER]: start probeDisaster
2020/12/12 19:08:01 [SERVER]: finish probeDisaster
2020/12/12 19:08:01 [SERVER]: start deductCostOfLiving
2020/12/12 19:08:01 [SERVER]: finish deductCostOfLiving
2020/12/12 19:08:01 [SERVER]: start updateIslandLivingStatus
2020/12/12 19:08:01 [SERVER]: finish updateIslandLivingStatus
2020/12/12 19:08:01 [SERVER]: finish endOfTurn
2020/12/12 19:08:01 [SERVER]: finish runTurn
2020/12/12 19:08:01 [SERVER]: start runTurn
2020/12/12 19:08:01 [SERVER]: TURN: 2, Season: 1
2020/12/12 19:08:01 [SERVER]: start updateIsland
2020/12/12 19:08:01 [Team2]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/12 19:08:01 [Team3]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/12 19:08:01 [Team4]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/12 19:08:01 [Team5]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/12 19:08:01 [Team6]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/12 19:08:01 [Team1]: Received game state update: {1 2 map[Team1:{90 Alive 0} Team2:{90 Alive 0} Team3:{90 Alive 0} Team4:{90 Alive 0} Team5:{90 Alive 0} Team6:{90 Alive 0}]}
2020/12/12 19:08:01 [SERVER]: finish updateIsland
2020/12/12 19:08:01 [SERVER]: start runOrgs
2020/12/12 19:08:01 [SERVER]: start runIITO
2020/12/12 19:08:01 [SERVER]: finish runIITO
2020/12/12 19:08:01 [SERVER]: start runIIFO
2020/12/12 19:08:01 [SERVER]: finish runIIFO
2020/12/12 19:08:01 [SERVER]: start runIITO
2020/12/12 19:08:01 [SERVER]: finish runIITO
2020/12/12 19:08:01 [SERVER]: finish runOrgs
2020/12/12 19:08:01 [SERVER]: start getEndOfTurnActions
2020/12/12 19:08:01 [Team1]: EndOfTurnActions
2020/12/12 19:08:01 [Team2]: EndOfTurnActions
2020/12/12 19:08:01 [Team3]: EndOfTurnActions
2020/12/12 19:08:01 [Team4]: EndOfTurnActions
2020/12/12 19:08:01 [Team5]: EndOfTurnActions
2020/12/12 19:08:01 [Team6]: EndOfTurnActions
2020/12/12 19:08:01 [SERVER]: finish getEndOfTurnActions
2020/12/12 19:08:01 [GAMESTATE]: start DispatchActions
2020/12/12 19:08:01 [GAMESTATE]: finish DispatchActions
2020/12/12 19:08:01 [SERVER]: start endOfTurn
2020/12/12 19:08:01 [SERVER]: start probeDisaster
2020/12/12 19:08:01 [SERVER]: finish probeDisaster
2020/12/12 19:08:01 [SERVER]: start deductCostOfLiving
2020/12/12 19:08:01 [SERVER]: finish deductCostOfLiving
2020/12/12 19:08:01 [SERVER]: start updateIslandLivingStatus
2020/12/12 19:08:01 [SERVER]: finish updateIslandLivingStatus
2020/12/12 19:08:01 [SERVER]: finish endOfTurn
2020/12/12 19:08:01 [SERVER]: finish runTurn
2020/12/12 19:08:01 [SERVER]: Max turns '2' reached or exceeded
===== START OF TURN 1 (END OF TURN 0) =====
common.GameState{Season:0x1, Turn:0x1, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:100, LifeStatu
s:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0
x0}, Team5:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:100, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 2 (END OF TURN 1) =====
common.GameState{Season:0x1, Turn:0x2, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:90, LifeStatus:
Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0},
 Team5:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:90, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}
===== START OF TURN 3 (END OF TURN 2) =====
common.GameState{Season:0x1, Turn:0x3, ClientInfos:map[shared.ClientID]common.ClientInfo{Team1:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team2:common.ClientInfo{Resources:80, LifeStatus:
Alive, CriticalConsecutiveTurnsCounter:0x0}, Team3:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team4:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0},
 Team5:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}, Team6:common.ClientInfo{Resources:80, LifeStatus:Alive, CriticalConsecutiveTurnsCounter:0x0}}}

```

### CODE REVIEW

**Review by lhl2617 — APPROVED** [review]
image

**Comment by lhl2617** [comment]
should merge main into that branch before pulling in btw, but this is fine.

File: internal/clients/team3/client.go [hunk 1]
```diff
@@ -3,42 +3,56 @@ package team3
...
+	evalResult, err := rules.BasicBooleanRuleEvaluator("Kinda Complicated Rule")
 	if err != nil {
 		panic(err.Error())
 	}
-	c.Logf("Rule Eval: %t", evalResult)
+	c.logf("Rule Eval: %t", evalResult)
 }
 
-func (c *client) GetID() common.ClientID {
+func (c *client) GetID() shared.ClientID {
```
Comment (Eirikalb) [inline internal/clients/team3/client.go:38]: This should be a merge conflict... the GetID method is implemented in the BaseClient now see #45. A feature addition should not change the individual client files at all following that PR, only the BaseClient

File: internal/clients/team3/client.go [hunk 2]
```diff
@@ -3,42 +3,56 @@ package team3
...
 	return c.id
 }
 
-// Logf is the client's logger that prepends logs with your ID. This makes
+// logf is the client's logger that prepends logs with your ID. This makes
 // it easier to read logs. DO NOT use other loggers that will mess logs up!
-func (c *client) Logf(format string, a ...interface{}) {
+func (c *client) logf(format string, a ...interface{}) {
```
Comment (Eirikalb) [inline internal/clients/team3/client.go:44]: This is also implemented in the BaseClient, so this method should not need to exist in the clients.

File: internal/clients/team3/client.go [hunk 3]
```diff
@@ -3,42 +3,56 @@ package team3
...
 	log.Printf("[%v]: %v", c.id, fmt.Sprintf(format, a...))
 }
+
```
Comment (Eirikalb) [inline internal/clients/team3/client.go:47]: These methods are implemented in the BaseClient, but will need to be implemented in the individual clients as well.

File: internal/clients/team4/README.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# Team 4: Cherry Pitt 🍒
```
Comment (Eirikalb) [inline internal/clients/team4/README.md:1]: Arents these files already in the main branch? I am confused as to what this diff actually shows now?

Comment (Eirikalb) [reply internal/clients/team4/README.md:1]: Ahh now I see, it is not going onto the main branch..... Me understands

**Review by Eirikalb — APPROVED** [review]
The rule thing LGTM. Before you merge to main make sure to address the comments I made about the BaseClient, I will go through the client better abit more on Monday I thought

**Comment by QFSW** [comment]
Agreed with LH that you should merge into your branch before PRing, that way you can run sim/tests and make sure that all would work on the result of the PR + produces a cleaner PR

**Pull request merged**